flink-end-to-end-tests/flink-end-to-end-tests-common/src/main/java/org/apache/flink/tests/util/flink/container/FlinkImageBuilder.java

```diff
@@ -0,0 +1,293 @@
+                                                        builder.copy(to.toString(), to.toString()));
+                                        builder.cmd(startupCommand);
+                                    });
+            filesToCopy.forEach((from, to) -> image.withFileFromPath(to.toString(), from));
+            return image;
+        } catch (Exception e) {
+            throw new ImageBuildException(
+                    String.format("Failed to build Flink image \"%s\"", finalImageName), e);
+        }
+    }
+
+    // ----------------------- Helper functions -----------------------
+
+    private void buildBaseImage(Path flinkDist) throws TimeoutException {
+        if (!baseImageExists()) {
+            LOG.info("Building Flink base image with flink-dist at {}", flinkDist);
+            new ImageFromDockerfile(FLINK_BASE_IMAGE_NAME)
+                    .withDockerfileFromBuilder(
+                            builder ->
+                                    builder.from("openjdk:" + getJavaVersionSuffix())
+                                            .copy("flink", "flink")
+                                            .build())
+                    .withFileFromPath("flink", flinkDist)
+                    .get(timeout.toMillis(), TimeUnit.MILLISECONDS);
+        }
+    }
+
+    private boolean baseImageExists() {
+        try {
+            DockerClientFactory.instance().client().inspectImageCmd(FLINK_BASE_IMAGE_NAME).exec();
+            return true;
+        } catch (NotFoundException e) {
+            return false;
+        }
+    }
+
+    private String getJavaVersionSuffix() {
+        if (this.javaVersion != null) {
+            return this.javaVersion;
+        } else {
+            String javaSpecVersion = System.getProperty("java.vm.specification.version");
+            LOG.info("Using JDK version {} of the current VM", javaSpecVersion);
+            switch (javaSpecVersion) {
+                case "1.8":
+                    return "8";
+                case "11":
+                    return "11";
+                default:
+                    throw new IllegalStateException("Unexpected Java version: " + javaSpecVersion);
+            }
+        }
+    }
+
+    private Path createTemporaryFlinkConfFile(Path tempDirectory) throws IOException {
+        // Load Flink configurations in flink-dist
+        final Configuration finalConfiguration =
+                GlobalConfiguration.loadConfiguration(
+                        flinkDist.resolve("conf").toAbsolutePath().toString());
+
+        if (this.conf != null) {
+            finalConfiguration.addAll(this.conf);
+        }
+
+        Path flinkConfFile = tempDirectory.resolve(GlobalConfiguration.FLINK_CONF_FILENAME);
+
+        Files.write(
+                flinkConfFile,
+                finalConfiguration.toMap().entrySet().stream()
+                        .map(entry -> entry.getKey() + ": " + entry.getValue())
+                        .collect(Collectors.toList()));
+
```

Comment:
This method looks a bit complicated can you add a docstring what you are trying to do here?